### ISSUE TITLE
[nixpkgs] Trying to set the "watch" option in the configuration file seems to be ignored

### ISSUE THREAD

**Issue by gautaz** [f3d-app/f3d#1726]
**Describe the bug**
I am most of the time using f3d with the `watch` option so I wanted to enable it through the configuration file.

**To Reproduce**
Here is the configuration file I am using:

```json
{
        ".*": {
                "ambient-occlusion": true,
                "axis": true,
                "camera-direction": "1,1,-1",
                "grid": true,
                "grid-absolute": true,
                "up": "+Z",
                "watch": true
        }
}
```

All options seem to be taken into account but the `watch` option.

**Expected behavior**
The [documentation about the configuration file](https://github.com/f3d-app/f3d/blob/master/doc/user/CONFIGURATION_FILE.md) states:

> The following options cannot be set via config file: help, version, readers-list, config, dry-run and input.

So I expected `watch` to be set via config file.

**System Information:**
 - OS: Linux 6.6.62 NixOS SMP PREEMPT_DYNAMIC Sun Nov 17 14:08:59 UTC 2024 x86_64 GNU/Linux
 - GPU and GPU driver: I have no idea, providing a command to retrieve this information would be helpful.

**F3D Information**
Paste the content of `f3d --version`:
```console
F3D 2.5.0

F3D - A fast and minimalist 3D viewer
Version: 2.5.0.
Build date: 1980-01-01 00:00:00.
Build system: Linux 64-bits.
Compiler: GNU 13.3.0.
External rendering module: ON.
Raytracing module: OFF.
VTK version: 9.2.6.
Copyright (C) 2019-2021 Kitware SAS.
Copyright (C) 2021-2024 Michael Migliore, Mathieu Westphal.
License BSD-3-Clause.
By Michael Migliore, Mathieu Westphal and Joachim Pouderoux.
```

**Additional context**

**Comment by mwestphal** [comment]
Hi @gautaz 

I'm unable to reproduce the issue with F3D 2.5.0 binary release on Linux.

Please run F3D with `--verbose` in order to figure out if your config file is being picked up correctly and the watch option being read as expected. On my side it looks like this:

```
using `global` config from /home/glow/dev/f3d/others/F3D-2.5.0-RC1-Linux-x86_64-raytracing/share/f3d/configs/config.d/10_global.json
+ animation-progress: true
+ anti-aliasing: true
+ axis: true
+ camera-direction: -1,-0.5,-1
+ filename: true
+ grid: true
+ hdri-ambient: true
+ light-intensity: 2
+ progress: true
+ scalars: 
+ tone-mapping: true
+ translucency-support: true
+ watch: true
========== Configuring engine ==========
Engine configured
========== Loading 3D file ==========
using `global` config from /home/glow/dev/f3d/others/F3D-2.5.0-RC1-Linux-x86_64-raytracing/share/f3d/configs/config.d/10_global.json
+ animation-progress: true
+ anti-aliasing: true
+ axis: true
+ camera-direction: -1,-0.5,-1
+ filename: true
+ grid: true
+ hdri-ambient: true
+ light-intensity: 2
+ progress: true
+ scalars: 
+ tone-mapping: true
+ translucency-support: true
+ watch: true

```

**Comment by gautaz** [comment]
Hello @mwestphal,

Thank you for taking some time to diagnose the issue (which might be between the keyboard and the chair).

Here is what `f3d --verbose` outputs on my side:

```console
❯ f3d --verbose
========== Initializing ==========
Loading plugin "assimp"
  Version: 1.0
  Description: Assimp support (version 5.4.3)
  Readers:
    Filmbox
    Collada
    AutoCAD DXF
    Object File Format
    DirectX File Format
    3D Manufacturing Format
Loading plugin "exodus"
  Version: 1.0
  Description: VTK Exodus support
  Readers:
    Exodus II
Loading plugin "native"
  Version: 1.0
  Description: Native VTK I/O support
  Readers:
    Autodesk 3D Studio
    CityGML
    DICOM
    GL Transmission Format
    MetaImage
    Nearly Raw Raster Data
    Wavefront OBJ
    Polygon
    Point Cloud
    Standard Triangle Language
    TIFF
    VRML
    VTK Legacy
    VTK XML UnstructuredGrid
    VTK XML PolyData
    VTK XML ImageData
    VTK XML RectangularGrid
    VTK XML StructuredGrid
    VTK XML MultiBlock
    3D Gaussian splats
Loading plugin "occt"
  Version: 1.0
  Description: OpenCASCADE support (version 7.8.1)
  Readers:
    STEP ISO 10303
    Initial Graphics Exchange Specification
    Open CASCADE BRep
Using config file /nix/store/c177mfm8bgjz5kkdcphvvjy2mmjfa7bv-source/homes/del/f3d.json
========== Configuring engine ==========
Engine configured
========== Loading 3D file ==========
No file to load provided.
========== Rendering ==========
Render done
```

And to confirm the content of the configuration file:

```console
❯ cat /nix/store/c177mfm8bgjz5kkdcphvvjy2mmjfa7bv-source/homes/del/f3d.json
{
        ".*": {
                "ambient-occlusion": true,
                "axis": true,
                "camera-direction": "1,1,-1",
                "grid": true,
                "grid-absolute": true,
                "up": "+Z",
                "watch": true
        }
}
```

As for the current behavior on my side, I am using `f3d` to display a 3MF model.
When the model is modified, nothing happens.
But when I press the UP key then the modifications are displayed.

Feel free to ask for any missing detail that might lead to an explanation of this behavior.

**Comment by gautaz** [comment]
And here is the output when viewing the 3MF file:

```console
❯ f3d --verbose models/bookstand.3mf                                                                                                                                                                                                                                                                                                                                                          
========== Initializing ==========
Loading plugin "assimp"
  Version: 1.0
  Description: Assimp support (version 5.4.3)
  Readers:
    Filmbox
    Collada
    AutoCAD DXF
    Object File Format
    DirectX File Format
    3D Manufacturing Format
Loading plugin "exodus"
  Version: 1.0
  Description: VTK Exodus support
  Readers:
    Exodus II
Loading plugin "native"
  Version: 1.0
  Description: Native VTK I/O support
  Readers:
    Autodesk 3D Studio
    CityGML
    DICOM
    GL Transmission Format
    MetaImage
    Nearly Raw Raster Data
    Wavefront OBJ
    Polygon
    Point Cloud
    Standard Triangle Language
    TIFF
    VRML
    VTK Legacy
    VTK XML UnstructuredGrid
    VTK XML PolyData
    VTK XML ImageData
    VTK XML RectangularGrid
    VTK XML StructuredGrid
    VTK XML MultiBlock
    3D Gaussian splats
Loading plugin "occt"
  Version: 1.0
  Description: OpenCASCADE support (version 7.8.1)
  Readers:
    STEP ISO 10303
    Initial Graphics Exchange Specification
    Open CASCADE BRep
Using config file /nix/store/c177mfm8bgjz5kkdcphvvjy2mmjfa7bv-source/homes/del/f3d.json
========== Configuring engine ==========
Engine configured
========== Loading 3D file ==========
using `.*` config from /nix/store/c177mfm8bgjz5kkdcphvvjy2mmjfa7bv-source/homes/del/f3d.json
+ ambient-occlusion: true
+ axis: true
+ camera-direction: 1,1,-1
+ grid: true
+ grid-absolute: true
+ up: +Z
+ watch: true
Found a reader for "/home/del/github.com/gautaz/cadcode/models/bookstand.3mf" : "3MF"
Loading 3D scene: /home/del/github.com/gautaz/cadcode/models/bookstand.3mf

No animation available in this file
No camera available in this file

Scene Graph:
------------
3MF
 Object_1


Scene bounding box: 0,120.754,13.3511,156.472,0,6

Camera position: 60.3769,-277.026,3
Camera focal point: 60.3769,84.9113,3
Camera view up: 0,0,1
Camera view angle: 15.5

Using grid unit square size = 10
Grid origin set to [60.3769, 84.9113, 0]


========== Rendering ==========
Render done
========== Loading 3D file ==========
using `.*` config from /nix/store/c177mfm8bgjz5kkdcphvvjy2mmjfa7bv-source/homes/del/f3d.json
+ ambient-occlusion: true
+ axis: true
+ camera-direction: 1,1,-1
+ grid: true
+ grid-absolute: true
+ up: +Z
+ watch: true
Found a reader for "/home/del/github.com/gautaz/cadcode/models/bookstand.3mf" : "3MF"
Loading 3D scene: /home/del/github.com/gautaz/cadcode/models/bookstand.3mf

No animation available in this file
No camera available in this file

Scene Graph:
------------
3MF
 Object_1


Scene bounding box: 0,120.754,13.3516,156.471,0,6

Camera position: 60.3769,-316.466,3
Camera focal point: 60.3769,84.9115,3
Camera view up: 0,0,1
Camera view angle: 25.4175

Using grid unit square size = 10
Grid origin set to [60.3769, 84.9115, 0]


========== Rendering ==========
Render done
```

**Comment by mwestphal** [comment]
> + watch: true

Afaics the option is correclty read from the config file, what happens in the terminal when you modifiy the file externally ?

**Comment by gautaz** [comment]
Here is a transcript with my actions:

```console
❯ f3d --verbose models/bookstand.3mf
========== Initializing ==========
Loading plugin "assimp"
  Version: 1.0
  Description: Assimp support (version 5.4.3)
  Readers:
    Filmbox
    Collada
    AutoCAD DXF
    Object File Format
    DirectX File Format
    3D Manufacturing Format
Loading plugin "exodus"
  Version: 1.0
  Description: VTK Exodus support
  Readers:
    Exodus II
Loading plugin "native"
  Version: 1.0
  Description: Native VTK I/O support
  Readers:
    Autodesk 3D Studio
    CityGML
    DICOM
    GL Transmission Format
    MetaImage
    Nearly Raw Raster Data
    Wavefront OBJ
    Polygon
    Point Cloud
    Standard Triangle Language
    TIFF
    VRML
    VTK Legacy
    VTK XML UnstructuredGrid
    VTK XML PolyData
    VTK XML ImageData
    VTK XML RectangularGrid
    VTK XML StructuredGrid
    VTK XML MultiBlock
    3D Gaussian splats
Loading plugin "occt"
  Version: 1.0
  Description: OpenCASCADE support (version 7.8.1)
  Readers:
    STEP ISO 10303
    Initial Graphics Exchange Specification
    Open CASCADE BRep
Using config file /nix/store/c177mfm8bgjz5kkdcphvvjy2mmjfa7bv-source/homes/del/f3d.json
========== Configuring engine ==========
Engine configured
========== Loading 3D file ==========
using `.*` config from /nix/store/c177mfm8bgjz5kkdcphvvjy2mmjfa7bv-source/homes/del/f3d.json
+ ambient-occlusion: true
+ axis: true
+ camera-direction: 1,1,-1
+ grid: true
+ grid-absolute: true
+ up: +Z
+ watch: true
Found a reader for "/home/del/github.com/gautaz/cadcode/models/bookstand.3mf" : "3MF"
Loading 3D scene: /home/del/github.com/gautaz/cadcode/models/bookstand.3mf

No animation available in this file
No camera available in this file

Scene Graph:
------------
3MF
 Object_1


Scene bounding box: 0,120.754,13.3516,156.471,0,6

Camera position: 60.3769,-277.025,3
Camera focal point: 60.3769,84.9115,3
Camera view up: 0,0,1
Camera view angle: 15.5

Using grid unit square size = 10
Grid origin set to [60.3769, 84.9115, 0]


========== Rendering ==========
Render done
```

At this point the file is displayed.
Then I operate the modification and nothing happens on the console.
To conclude, I finally press the up key which updates the view and displays the following on the console:

```console
========== Loading 3D file ==========
using `.*` config from /nix/store/c177mfm8bgjz5kkdcphvvjy2mmjfa7bv-source/homes/del/f3d.json
+ ambient-occlusion: true
+ axis: true
+ camera-direction: 1,1,-1
+ grid: true
+ grid-absolute: true
+ up: +Z
+ watch: true
Found a reader for "/home/del/github.com/gautaz/cadcode/models/bookstand.3mf" : "3MF"
Loading 3D scene: /home/del/github.com/gautaz/cadcode/models/bookstand.3mf

No animation available in this file
No camera available in this file

Scene Graph:
------------
3MF
 Object_1


Scene bounding box: 0,120.754,13.3511,156.472,0,6

Camera position: 60.3769,-277.026,3
Camera focal point: 60.3769,84.9113,3
Camera view up: 0,0,1
Camera view angle: 14.3876

Using grid unit square size = 10
Grid origin set to [60.3769, 84.9113, 0]


========== Rendering ==========
Render done
```

**Comment by mwestphal** [comment]
I really have no idea why this would happen only for you.

Can you try with the last nightly ?

https://github.com/f3d-app/f3d/releases/tag/nightly

Anything specific about your setup ? desktop env ? file manager ?

**Comment by gautaz** [comment]
As for my setup, I use [XMonad](https://xmonad.org/) as my window manager.
There is no desktop env per se as I like keeping things minimal (so no gnome/kde or anything like that).
I also don't use any file manager as I am using a terminal most of the time.

I will try the last nightly to see if it fixes the issue (I will need some more spare time though).
What is really strange is that setting the `--watch` option on the command line is working as expected so if the issue was due to the environment, I am wondering how the `f3d` command line parameters would be mitigating this...

**Comment by mwestphal** [comment]
Yes, I dont see a code path where that would matter tbh. Please let me know how the nightly goes.

**Comment by gautaz** [comment]
I have just tried with a version recompiled from master and it does not reproduce the issue.

Basically:
```console
❯ ./bin/f3d \
--config /nix/store/c177mfm8bgjz5kkdcphvvjy2mmjfa7bv-source/homes/del/f3d.json \
--verbose ../../../gautaz/cadcode/models/bookstand.3mf
```
works as intended.

So this seems to have something to do with the precompiled version from nixpkgs.
I need to investigate this a little further.

**Comment by mwestphal** [comment]
> So this seems to have something to do with the precompiled version from nixpkgs.

Could you try with our binary release from https://f3d.app/doc/user/INSTALLATION.html ?

Just to make sure it is indeed a nixpkgs problem, and not a larger packaging problem

**Comment by gautaz** [comment]
> Could you try with our binary release from https://f3d.app/doc/user/INSTALLATION.html ?

I would like, but NixOS poses some interesting challenges when one is wanting to execute programs from outside the Nix store.
As an example, this is what happens with the executable coming from the Linux (portable) tar archive:

```console
❯ ./f3d --help
Could not start dynamically linked executable: ./f3d
NixOS cannot run dynamically linked executables intended for generic
linux environments out of the box. For more information, see:
https://nix.dev/permalink/stub-ld
```

**Comment by mwestphal** [comment]
I had no idea!

I'll assume this is a nixpks specific issue then, please let us know how it goes and if F3D can do anything to help!

**Comment by pbsds** [comment]
I can reproduce the issue with both the nixpkgs-provided binary and the portable binary from https://f3d.app/doc/user/INSTALLATION.html (v2.5.0). It works with `--watch` but not when only set in the config file. I used the same config file as gautaz, and even made sure it was a file and not a symlink to the nix store.

<details>
<summary>Details</summary>

```
========== Initializing ==========
Loading plugin "alembic"
  Version: 1.0
  Description: Alembic support (version 1.8.5)
  Readers:
    Alembic
Loading plugin "assimp"
  Version: 1.0
  Description: Assimp support (version 5.4.0)
  Readers:
    Filmbox
    Collada
    AutoCAD DXF
    Object File Format
    DirectX File Format
    3D Manufacturing Format
Loading plugin "draco"
  Version: 1.0
  Description: Draco support (version 1.5.6)
  Readers:
    Draco
    GL Transmission Format
Loading plugin "exodus"
  Version: 1.0
  Description: VTK Exodus support
  Readers:
    Exodus II
Loading plugin "native"
  Version: 1.0
  Description: Native VTK I/O support
  Readers:
    Autodesk 3D Studio
    CityGML
    DICOM
    GL Transmission Format
    MetaImage
    Nearly Raw Raster Data
    Wavefront OBJ
    Polygon
    Point Cloud
    Standard Triangle Language
    TIFF
    VRML
    VTK Legacy
    VTK XML UnstructuredGrid
    VTK XML PolyData
    VTK XML ImageData
    VTK XML RectangularGrid
    VTK XML StructuredGrid
    VTK XML MultiBlock
    3D Gaussian splats
Loading plugin "occt"
  Version: 1.0
  Description: OpenCASCADE support (version 7.7.2)
  Readers:
    STEP ISO 10303
    Initial Graphics Exchange Specification
    Open CASCADE BRep
Loading plugin "usd"
  Version: 1.0
  Description: USD support
  Readers:
    Universal Scene Descriptor
Loading plugin "vdb"
  Version: 1.0
  Description: VTK OpenVDB support
  Readers:
    VDB
Using config directory /home/pbsds/Downloads/F3D-2.5.0-Linux-x86_64-raytracing/share/f3d/configs/config.d
Using config file /home/pbsds/.config/f3d/config.json
using `global` config from /home/pbsds/Downloads/F3D-2.5.0-Linux-x86_64-raytracing/share/f3d/configs/config.d/10_global.json
+ animation-progress: true
+ anti-aliasing: true
+ axis: true
+ camera-direction: -1,-0.5,-1
+ filename: true
+ grid: true
+ hdri-ambient: true
+ progress: true
+ scalars: 
+ tone-mapping: true
+ translucency-support: true
========== Configuring engine ==========
Engine configured
========== Loading 3D file ==========
using `global` config from /home/pbsds/Downloads/F3D-2.5.0-Linux-x86_64-raytracing/share/f3d/configs/config.d/10_global.json
+ animation-progress: true
+ anti-aliasing: true
+ axis: true
+ camera-direction: -1,-0.5,-1
+ filename: true
+ grid: true
+ hdri-ambient: true
+ progress: true
+ scalars: 
+ tone-mapping: true
+ translucency-support: true
using `.*(ply)` config from /home/pbsds/Downloads/F3D-2.5.0-Linux-x86_64-raytracing/share/f3d/configs/config.d/10_native.json
+ comp: -2
= translucency-support: true
using `.*` config from /home/pbsds/.config/f3d/config.json
+ ambient-occlusion: true
= axis: true
- camera-direction: -1,-0.5,-1
+ camera-direction: 1,1,-1
= grid: true
+ grid-absolute: true
+ up: +Z
+ watch: true
Found a reader for "/home/pbsds/Documents/3d/stanford-bunny.ply" : "PLYReader"
Loading 3D geometry: /home/pbsds/Documents/3d/stanford-bunny.ply

No animation available in this file
No camera available in this file

=== stanford-bunny.ply ===
Number of points: 35947
Number of polygons: 69451
Number of lines: 0
Number of vertices: 0
0 point data array(s):
0 cell data array(s):
0 field data array(s):


No array to color with
Not coloring

Scene bounding box: -0.0946899,0.0610091,0.0329874,0.187321,-0.0618736,0.0587997

Camera position: -0.0168404,-0.373284,-0.00153695
Camera focal point: -0.0168404,0.110154,-0.00153695
Camera view up: 0,0,1
Camera view angle: 18.4444

Using grid unit square size = 0.01
Grid origin set to [-0.0168404, 0.110154, 0]


No array to color with
========== Rendering ==========
No array to color with
Render done
```

</details>

_(I got the portable binary to run on nixos using `nix-ld` and invoking it as `LD_LIBRARY_PATH="$(realpath lib)" bin/f3d`.)_

**Comment by pbsds** [comment]
I tried running f3d without `--watch` with `strace`, and saw no action as i modified the 3d model.

**Comment by mwestphal** [comment]
Here is what I did to try to reproduce the isssue which **did NOT reproduce** it:

 - Download https://github.com/f3d-app/f3d/releases/download/v2.5.0/F3D-2.5.0-Linux-x86_64-raytracing.tar.xz
 - Extract it
 - edit `share/f3d/configs/config.d/10_global.json` and add `"watch": true,` like this:
 
 ```
 {
  "global":
  {
    "axis": true,
    "tone-mapping": true,
    "grid": true,
    "progress": true,
    "anti-aliasing": true,
    "filename": true,
    "scalars": "", 
    "camera-direction": "-1,-0.5,-1",
    "hdri-ambient": true,
    "translucency-support": true,
    "watch": true,
    "animation-progress": true
  }
}
 ```
 
 - run ./bin/f3d cow.vtp
 - in another terminal `cp cowlow.vtp cow.vtp`
 
 F3D updates as expected.

Please try to follow these steps exactly and adds only the minimum of changes requiring to make these steps makes sense on nix, and let me know how it goes. I want to get to the bottom of this.

**Comment by pbsds** [comment]
`share/f3d/configs/config.d/10_global.json` works as if i had added `--watch`, the issue must be with the one in `$XDG_CONFIG_HOME/f3d/config.json`

**Comment by mwestphal** [comment]
Thats progress! Let me try by relying on XDG_CONFIG_HOME

**Comment by pbsds** [comment]
It works in `XDG_CONFiG_HOME` when using `"globel"`, but not when using `".*"`

**Comment by mwestphal** [comment]
More progress!

Well Im glad to say this issue is already fixed, and I should have caught it on the first post.

Indeed, this option MUST be set in the "global" part. However we removed that limitation in the development version of F3D, but our documentation (https://f3d.app) is about the developement version of F3D, not our last release, this is why this limitation is not mentionned there (anymore).

related issue: https://github.com/f3d-app/f3d/issues/845

TLDR: Set `watch` option in `global` config for 2.5.0 version of F3D.
